utils/src/main/java/software/amazon/awssdk/utils/cache/NonBlocking.java

```diff
@@ -37,40 +41,107 @@
+                                                                     .build());
+
+    /**
+     * Threads used to do the actual work of refreshing the credentials (because the cached supplier might block, so we don't
+     * want the work to be done by a small thread pool). This executor is created as unbounded, but in reality it is limited by
+     * the {@link #MAX_CONCURRENT_REFRESHES} via {@link #BACKGROUND_REFRESH_LEASE}.
+     */
+    private static final ThreadPoolExecutor EXECUTOR =
+        new ThreadPoolExecutor(0,
+                               Integer.MAX_VALUE,
+                               DEFAULT_REFRESH_FREQUENCY.toMillis() + DEFAULT_REFRESH_FREQUENCY_JITTER.toMillis() + 5_000,
+                               MILLISECONDS,
+                               new SynchronousQueue<>(),
+                               new ThreadFactoryBuilder().daemonThreads(true).build());
+    /**
+     * A set of leases used to prevent concurrent refreshes beyond the limit described in {@link #MAX_CONCURRENT_REFRESHES}. If
+     * a lease cannot be acquired, the refresh is skipped.
+     */
+    private static final Semaphore BACKGROUND_REFRESH_LEASE = new Semaphore(MAX_CONCURRENT_REFRESHES);
+
+    /**
+     * An incrementing number, used to uniquely identify an instance of NonBlocking in the {@link #asyncThreadName}.
+     */
+    private static final AtomicLong THREAD_NUMBER = new AtomicLong(0);
+
     /**
      * Whether we are currently refreshing the supplier. This is used to make sure only one caller is blocking at a time.
      */
     private final AtomicBoolean currentlyRefreshing = new AtomicBoolean(false);
 
+    /**
+     * Name of the thread refreshing the cache for this strategy.
+     */
+    private final String asyncThreadName;
+
     /**
      * How frequently to automatically refresh the supplier in the background.
      */
     private final Duration asyncRefreshFrequency;
 
     /**
-     * Single threaded executor to asynchronous refresh the value.
+     * Whether this strategy has been shutdown (and should stop doing background refreshes)
      */
-    private final ScheduledExecutorService executor;
+    private volatile boolean shutdown = false;
 
     /**
      * Create a non-blocking prefetch strategy that uses the provided value for the name of the background thread that will be
      * performing the update.
      */
     public NonBlocking(String asyncThreadName) {
-        this(asyncThreadName, Duration.ofMinutes(1));
+        this(asyncThreadName, defaultRefreshFrequency());
+    }
+
+    private static Duration defaultRefreshFrequency() {
+        // We jitter the default refresh frequency with each instance, so that objects created at the same time will not all be
+        // refreshing at the exact same times.
+        int jitter = Math.toIntExact(DEFAULT_REFRESH_FREQUENCY_JITTER.toMillis());
+        long asyncRefreshFrequency = DEFAULT_REFRESH_FREQUENCY.toMillis() +
```

Comment:
With SDK 2.17.214, as expected, cred providers do their async refresh all at the same time (when created at the same time) - and in separate thread pools.

With this PR, cred providers do their async refresh over an approx 1m window (when created at the same time) - and in the newly introduced shared thread pool.  However, the time distributions aren't uniform (some bucketing due to the below).

The time distributions are hard to get one's head around due to:
  -jittered time for async wakeup tasks;
  -prefetch time for STS cred provider;

Perhaps the prefetch strategy could be simplified by utilizing the actual cred expiration and the configured prefetch time (re-interpreted as a prefetch window).  This would avoid the "poll every 60-70s and then check prefetch" (which starts to align tasks), allowing for a clean distribution of tasks over a prefetch window.

Reply:
Yeah, it's a bit confusing to wrap ones head around since most of the time it's just a timer that hits the cache or a timer that kicks off a prefetch.

It would be nice to be able to take the prefetch window into consideration, but that would be an API change I was hoping to avoid. I suppose it's not likely to be problematic to make that change so that the prefetch strategies can be aware of the prefetch times.

Reply:
@cloudshiftchris I've taken a shot at it. After we have a successful fetch, we schedule a hit against the cached supplier for a time after the prefetch time. We randomly jitter that time a bit between the prefetch time and the stale time.